### PR TITLE
circles tool

### PR DESCRIPTION
**Pull request by xmonader**
wip for https://github.com/threefoldtech/circles_reporting_tool/blob/master/specs/

### CODE REVIEW

File: docs/wiki/tutorials/taiga_client.md [hunk 1]
```diff
@@ -1,80 +1,141 @@
 # Taiga client
 
-This client is used to interact with Taiga API.
-
 ## Initialization
 
 Using username and  password:
-> client = j.clients.taiga.new('test', host="https://staging.circles.threefold.me/", username='7mada', password='123456')
-
+```
```
Comment (rkhamis) [inline docs/wiki/tutorials/taiga_client.md:6]: is 7amada really going to be part of the official documentation? :)

Comment (xmonader) [reply docs/wiki/tutorials/taiga_client.md:6]: updated to admin ^_^

File: jumpscale/clients/taiga/models.py [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+
+    @property
+    def url(self):
+        # https://circles.threefold.me/project/despiegk-tftech-software/issue/286
+        return f"{self._client.host}/project/{self.project_extra_info.get('slug')}/issue/{self.id}"
+
+    @property
+    def as_md(self):
+        TEMPLATE = """
+- **Subject:** [{{issue.subject}}]({{issue.url}})
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/models.py:30]: maybe if we use `dedent` here to align the indentation

File: jumpscale/clients/taiga/models.py [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+
+    @property
+    def url(self):
+        # https://circles.threefold.me/project/despiegk-tftech-software/us/214
+        return f"{self._client.host}/project/{self.project_extra_info.get('slug')}/us/{self.id}"
+
+    @property
+    def as_md(self):
+        TEMPLATE = """
+- **Subject:** [{{story.subject}}]({{story.url}})
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/models.py:61]: maybe if we use `dedent` here to align the indentation

File: jumpscale/clients/taiga/models.py [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+    def circles(self):
+        return self.get_circles()
+
+    def __dir__(self):
+        return dir(self._original_object) + ["as_md", "issues", "stories", "get_circles", "get_issues", "get_stories"]
+
+    @property
+    def as_md(self):
+
+        TEMPLATE = """
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/models.py:147]: maybe if we use `dedent` here to align the indentation

File: jumpscale/clients/taiga/models.py [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+    def __getattr__(self, attr):
+        return getattr(self._original_object, attr)
+
+    def __str__(self):
+        return f"<Circle {self._original_object}>"
+
+    @property
+    def as_md(self):
+
+        TEMPLATE = """
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/models.py:286]: maybe if we use `dedent` here to align the indentation

File: jumpscale/clients/taiga/taiga.py [hunk 6]
```diff
@@ -87,6 +162,9 @@
 from collections import defaultdict
 import gevent
 
+# from .models import User, Story, Issue, Task, Project, Circle
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:165]: we can clean unused comment

File: jumpscale/clients/taiga/taiga.py [hunk 7]
```diff
@@ -205,10 +297,10 @@ def get_user_stories(self, username):
         user_stories = self.api.user_stories.list(assigned_to=user_id)
         user_stories = []
         for user_story in user_stories:
-            user_story.project = self._get_project(user_story.project)
-            user_story.milestone = self._get_milestone(user_story.milestone)
+            # user_story.project = self._get_project(user_story.project)
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:300]: we can clean unused comment

File: jumpscale/clients/taiga/taiga.py [hunk 8]
```diff
@@ -221,14 +313,14 @@ def get_user_tasks(self, username):
         user_tasks = self.api.tasks.list(assigned_to=user_id)
         user_tasks = []
         for user_task in user_tasks:
-            user_task.project = self._get_project(user_task.project)
-            user_task.milestone = self._get_milestone(user_task.milestone)
+            # user_task.project = self._get_project(user_task.project)
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:316]: we can clean unused comment

File: jumpscale/clients/taiga/taiga.py [hunk 9]
```diff
@@ -331,279 +424,385 @@ def list_all_milestones(self):
...
-            if with_description:
-                text += f"  - **Description:** \n```\n{issue.get('description','unknown')}\n``` \n"
-        return text
-
-    def __render_issues_with_subjects(self, text="", issues=None):
-        """Render issues with subject only to be written to markdown file
+        return CircleIssue(self, self.api.issues.get(issue_id))
+
+    def _resolve_object(self, obj):
+        # print("resolving obj ", obj)
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:454]: we can clean unused comment

File: jumpscale/clients/taiga/taiga.py [hunk 10]
```diff
@@ -331,279 +424,385 @@ def list_all_milestones(self):
...
+        }
+        newobj = copy.deepcopy(obj)
+        for k in dir(newobj):
+            v = getattr(newobj, k)
+            if isinstance(v, int) or isinstance(v, list) and v and isinstance(v[0], int):
+                if k in resolvers:
+                    resolved = None
+                    resolver = resolvers[k]
+                    try:
+                        # print(f"resolver {resolver}")
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:475]: we can clean unused comment

File: jumpscale/clients/taiga/taiga.py [hunk 11]
```diff
@@ -331,279 +424,385 @@ def list_all_milestones(self):
...
+                        copied_v = copy.deepcopy(v)
+                        resolved = lambda: resolver(copied_v)
+
+                        if isinstance(v, list):
+                            # print(f"setting {k} to {resolved}")
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:480]: we can clean unused comment

File: jumpscale/clients/taiga/taiga.py [hunk 12]
```diff
@@ -331,279 +424,385 @@ def list_all_milestones(self):
...
+                            setattr(newobj, f"{k}_objects", resolved)
+                        else:
+                            # print(f"setting {k} to {resolved}")
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:483]: we can clean unused comment

File: jumpscale/clients/taiga/taiga.py [hunk 13]
```diff
@@ -331,279 +424,385 @@ def list_all_milestones(self):
...
+                            setattr(newobj, f"{k}_object", resolved)
+
+                    except Exception as e:
+                        import traceback
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:487]: perhaps we can move tracebook to top

File: jumpscale/clients/taiga/taiga.py [hunk 14]
```diff
@@ -331,279 +424,385 @@ def list_all_milestones(self):
...
+
+                        traceback.print_exc()
+
+                        print(f"error {e}")
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:491]: maybe we can replace print by logging

File: jumpscale/clients/taiga/taiga.py [hunk 15]
```diff
@@ -331,279 +424,385 @@ def list_all_milestones(self):
...
+
+        type_ = type_.upper()
+        project_name = f"{type_}_{name}"
+        p = self.api.projects.create(project_name, description=description)
+        for t in tasks_statuses:
+            try:
+                p.add_task_status(t)
+            except Exception as e:
+                # check if duplicated
+                print(f"skipping task {t} {e}")
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:543]: maybe we can replace print by logging

File: jumpscale/clients/taiga/taiga.py [hunk 16]
```diff
@@ -331,279 +424,385 @@ def list_all_milestones(self):
...
+
+        for t in priorities:
+            try:
+                p.add_priority(t)
+            except Exception as e:
+                # check if duplicated
+                print(f"skipping prio {t} {e}")
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:550]: maybe we can replace print by logging

File: jumpscale/clients/taiga/taiga.py [hunk 17]
```diff
@@ -331,279 +424,385 @@ def list_all_milestones(self):
...
+
+        for t in severities:
+            try:
+                p.add_severity(t)
+            except Exception as e:
+                # check if duplicated
+                print(f"skipping sever {t} {e}")
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:557]: maybe we can replace print by logging

File: jumpscale/clients/taiga/taiga.py [hunk 18]
```diff
@@ -331,279 +424,385 @@ def list_all_milestones(self):
...
+
+        for t in issues_statuses:
+            try:
+                p.add_issue_status(t)
+            except Exception as e:
+                # check if duplicated
+                print(f"skipping status {t} {e}")
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:564]: maybe we can replace print by logging

File: jumpscale/clients/taiga/taiga.py [hunk 19]
```diff
@@ -331,279 +424,385 @@ def list_all_milestones(self):
...
+
+        for t in user_stories_statuses:
+            try:
+                p.add_user_story_status(t)
+            except Exception as e:
+                # check if duplicated
+                print(f"skipping user status {t} {e}")
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:571]: maybe we can replace print by logging

File: jumpscale/clients/taiga/taiga.py [hunk 20]
```diff
@@ -331,279 +424,385 @@ def list_all_milestones(self):
...
+
+        for t in issues_types:
+            try:
+                p.add_issue_type(t)
+            except Exception as e:
+                # check if duplicated
+                print(f"skipping issue type {t} {e}")
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:578]: maybe we can replace print by logging

File: jumpscale/clients/taiga/taiga.py [hunk 21]
```diff
@@ -331,279 +424,385 @@ def list_all_milestones(self):
...
+
+        # list(map(p.add_task_status, tasks_statuses))
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:580]: cleaning unused code

File: jumpscale/clients/taiga/taiga.py [hunk 22]
```diff
@@ -331,279 +424,385 @@ def list_all_milestones(self):
...
-            # Sort the issues per group
-            sorted_list = sorted(projects_per_owner, key=lambda x: x.get("created_date"), reverse=True)
-            self.text = self.__render_projects(text=self.text, projects=sorted_list, with_details=with_details)
+        j.sals.fs.mkdirs(path)
+        circles = self.list_all_projects()
 
-    def export_issues_per_project(self, path="/tmp/issues_per_project.md", with_details=False):
-        """Export issues per project in a markdown file
+        def write_md_for_circle(circle):
+            # print(f"Writing {circle}")
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:741]: we can remove unused comment

File: jumpscale/clients/taiga/taiga.py [hunk 23]
```diff
@@ -331,279 +424,385 @@ def list_all_milestones(self):
...
-        self.text = f"""
-### {selected_user_name} Issues \n
-"""
-        self.map_render_issues_per_user(user_id, with_description)
+        users_mdpath = j.sals.fs.join_paths(path, "users.md")
+        users_mdcontent = "# users\n"
 
-        j.sals.fs.write_file(path=path, data=self.text)
+        def write_md_for_user(user):
+            # print(f"Writing {user}")
```
Comment (waleedhammam) [inline jumpscale/clients/taiga/taiga.py:779]: we can clean the unused comment

**Review by waleedhammam — APPROVED** [review]

File: jumpscale/clients/taiga/taiga.py [hunk 24]
```diff
@@ -192,7 +297,7 @@ def get_circles_issues(self, project_id):
             issue.priority = self._get_priority(issue.priority)
             issue.assignee = self._get_assignee(issue.assigned_to)
             issue.status = self._get_issue_status(issue.status)
-            circle_issues.append(issue)
+            circle_issues.append(self._resolve_objectissue)
```
Comment (abom) [inline jumpscale/clients/taiga/taiga.py:300]: what does this variable reference here?

File: jumpscale/clients/taiga/taiga.py [hunk 25]
```diff
@@ -205,10 +310,10 @@ def get_user_stories(self, username):
         user_stories = self.api.user_stories.list(assigned_to=user_id)
         user_stories = []
         for user_story in user_stories:
-            user_story.project = self._get_project(user_story.project)
-            user_story.milestone = self._get_milestone(user_story.milestone)
+            # user_story.project = self._get_project(user_story.project)
+            # user_story.milestone = self._get_milestone(user_story.milestone)
             user_story.status = self._get_user_stories_status(user_story.status)
-            user_stories.append(user_story)
+            user_stories.append(self._resolve_objectuser_story)
```
Comment (abom) [inline jumpscale/clients/taiga/taiga.py:316]: what does this variable reference here? _resolve_objectuser_story

**Review by abom — APPROVED** [review]

**Pull request merged**